### PR TITLE
feat(filter-menu): add header with search

### LINKED ISSUE #2329: filter-menu: add search header
# Enhancement Request

## Description

Add search header option to the `filter-menu` inline with Skin (https://github.com/eBay/skin/issues/2451) and allow filtering of the filter options.

### PR DESCRIPTION
**Pull request by LuLaValva**
- Resolves #2329 

## Description

- Add search header

## Context

- @ianmcburnie and I decided offline that filtering should happen _outside_ of eBayUI, because:
  1. There are likely multiple types of custom filtering (i. e. case sensitive, beginning of term vs within term)
  2. Simple cases that _would_ likely be a good candidate for eBayUI to do the filtering should likely not even have a search filter to begin with
  3. Network requests are impossible to handle in eBayUI

## Screenshots

image

image

### CODE REVIEW

File: src/components/ebay-filter-menu/component.ts [hunk 1]
```diff
@@ -56,6 +63,16 @@ export default class extends MenuUtils<Input, MenuState> {
         this.lastTabIndexPosition = 0;
     }
 
+    handleSearch(e: InputEvent) {
+        this.state.searchTerm = (e.target as HTMLInputElement).value;
+        this.emit("search-change", this.state.searchTerm);
```
Comment (agliga) [inline src/components/ebay-filter-menu/component.ts:68]: Do we want to add an option to filter the results? Like we have with combobox

Comment (agliga) [reply src/components/ebay-filter-menu/component.ts:68]: We can also iterate on this if needed

File: src/components/ebay-filter-menu/examples/with-search.marko
Comment (ArtBlue) [inline src/components/ebay-filter-menu/examples/with-search.marko:1]: A couple of things here...

(1) I think there's a bug here with more than one filtering.

* user enters search term
* user selects an item
* user clears search

The initial selection is reset/removed. It should be retained.

If a user is trying to find various items and doing multiple filtering, the prior selections should be maintained until manually removed to allow for more than one filtering and selection.

(2) A slight UX gap.

For longer lists, once a selection is made and context switched to other parts of the list via scrolling or filtering, the user does not have an easy way to reference the list of their current selections.

**Review by ArtBlue — CHANGES_REQUESTED** [review]
Looks good overall. One minor bug and a UX gap that we may not be able to do anything at the moment. I just wanted to document it.

File: src/components/ebay-filter-menu/component.ts (hunk 1 above)
Comment (LuLaValva) [reply src/components/ebay-filter-menu/component.ts:68]: I think it's overengineering to add our own filtering option, see the issue description for the results of a discussion with @ianmcburnie

File: src/components/ebay-filter-menu/examples/with-search.marko
Comment (LuLaValva) [reply src/components/ebay-filter-menu/examples/with-search.marko:1]: Good catch @ArtBlue on (1), I just committed a fix.

File: src/components/ebay-filter-menu/examples/with-search.marko [hunk 2]
```diff
@@ -1,31 +1,47 @@
...
     onCreate() {
         this.state = {
-            searchTerm: "",
+            checkedCodes: {},
+            visibleCountries: countries,
         };
     }
 
     handleSearchChange(value: string) {
-        this.state.searchTerm = value.toLowerCase();
```
[the remark this answers is not in the document]
Comment (agliga) [reply src/components/ebay-filter-menu/examples/with-search.marko]: Could there be a usecase where the developer of this component wants to pass in and sync up the state of the search
I'm thinking of page navigations or such. 
Generally I think its a good idea to emit the value of the search and then update a parents state and pass that in as input.

[the remark this answers is not in the document]
Comment (LuLaValva) [reply src/components/ebay-filter-menu/examples/with-search.marko]: I can't think of a usecase, but I guess there's no harm in adding it. I just pushed a commit that includes this functionality.

File: src/components/ebay-filter-menu/examples/with-search.marko
Comment (ArtBlue) [reply src/components/ebay-filter-menu/examples/with-search.marko:1]: @LuLaValva , looks awesome!

**Review by ArtBlue — APPROVED** [review]
Looks great! Let's just create an issue for that feature we can iterate on in the future.

**Pull request merged**